numpy/_core/tests/test_stringdtype.py

```diff
@@ -0,0 +1,1374 @@
+        pass
+
+    if is_nan or bool_errors or is_str:
+        for res in [arr * other, other * arr]:
+            np.testing.assert_array_equal(res[:-1], result)
+            if not is_str:
+                assert res[-1] is dtype.na_object
+            else:
+                try:
+                    assert res[-1] == dtype.na_object * other[-1]
+                except (IndexError, TypeError):
+                    assert res[-1] == dtype.na_object * other
+    else:
+        with pytest.raises(TypeError):
+            arr * other
+        with pytest.raises(TypeError):
+            other * arr
+
+
+def test_create_with_na(dtype):
+    if not hasattr(dtype, "na_object"):
+        pytest.skip("does not have an na object")
+    na_val = dtype.na_object
+    string_list = ["hello", na_val, "world"]
+    arr = np.array(string_list, dtype=dtype)
+    assert str(arr) == "[" + " ".join([repr(s) for s in string_list]) + "]"
+    assert arr[1] is dtype.na_object
+
+
+def test_datetime_cast(dtype):
+    a = np.array(
+        [
+            np.datetime64("1923-04-14T12:43:12"),
+            np.datetime64("1994-06-21T14:43:15"),
+            np.datetime64("2001-10-15T04:10:32"),
+            np.datetime64("NaT"),
+            np.datetime64("1995-11-25T16:02:16"),
+            np.datetime64("2005-01-04T03:14:12"),
+            np.datetime64("2041-12-03T14:05:03"),
+        ]
+    )
+
+    has_na = hasattr(dtype, "na_object")
+    is_str = isinstance(getattr(dtype, "na_object", None), str)
+
+    if not has_na or is_str:
+        a = np.delete(a, 3)
+
+    sa = a.astype(dtype)
+    ra = sa.astype(a.dtype)
+
+    if has_na and not is_str:
+        assert sa[3] is dtype.na_object
+        assert np.isnat(ra[3])
+
+    np.testing.assert_array_equal(a, ra)
+
+    if has_na and not is_str:
+        # don't worry about comparing how NaT is converted
+        sa = np.delete(sa, 3)
+        a = np.delete(a, 3)
+
+    np.testing.assert_array_equal(sa, a.astype("U"))
+
+
+def test_null_roundtripping(dtype):
+    data = ["hello\0world", "ABC\0DEF\0\0"]
+    arr = np.array(data, dtype=dtype)
+    assert data[0] == arr[0]
+    assert data[1] == arr[1]
+
+
+def test_string_too_large_error():
+    arr = np.array(["a", "b", "c"], dtype=StringDType())
+    with pytest.raises(MemoryError):
+        arr * (2**63 - 2)
+
+
+def test_growing_strings(dtype):
+    # growing a string leads to a heap allocation, this tests to make sure
+    # we do that bookeeping correctly for all possible starting cases
+    data = [
+        "hello",  # a short string
+        "abcdefghijklmnopqestuvwxyz",  # a medium heap-allocated string
+        "hello" * 200,  # a long heap-allocated string
+    ]
+
+    arr = np.array(data, dtype=dtype)
+    uarr = np.array(data, dtype=str)
+
+    for _ in range(5):
+        arr = arr + arr
+        uarr = uarr + uarr
+
+    np.testing.assert_array_equal(arr, uarr)
+
+
+def test_threaded_access_and_mutation(dtype, random_string_list):
```

Comment:
Should this test be marked `slow`?

Reply:
It's no longer slow after reducing the size of the `random_string_list`, so I'm not marking it as such.